src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+        exception will be raised.
+
+        The name of the loop index is taken from the PSyclone
+        configuration file if a name exists for the particular array
+        index, otherwise a new name is generated.
+
+        :param node: an array index.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        '''
+        self.validate(node)
+
+        array_index = node.position
+        array_reference = node.parent
+        assignment = array_reference.parent
+        symbol_table = node.scope.symbol_table
+
+        node_copy = node.copy()
+
+        # See if there is any configuration information for this array index
+        loop_type_order = Config.get().api_conf("nemo").get_index_order()
+        # TODO: Add tests in get_loop_type_data() to make sure values
+        # are strings that represent an integer or a valid variable
+        # name, e.g. 1a should not be allowed. See issue #1035
+        loop_type_data = Config.get().api_conf("nemo").get_loop_type_data()
+        try:
+            loop_type = loop_type_order[array_index]
+            loop_type_info = loop_type_data[loop_type]
+            loop_variable_name = loop_type_info['var']
+        except IndexError:
+            loop_variable_name = symbol_table.next_available_name("idx")
+
+        # Work out where to add the new loop (at 'location'), as there
+        # may be existing loops that should be inside this one.
+        symbols = [ref.symbol for ref in array_reference.children[:array_index]
+                   if isinstance(ref, Reference)]
+        location = assignment
+        nloops = 0
+        while (isinstance(location.parent, Schedule) and
+               isinstance(location.parent.parent, (Loop, InlinedKern)) and
+               nloops < len(symbols)):
+            if isinstance(location.parent.parent, Loop):
+                nloops += 1
+            location = location.parent.parent
+
+        # Look up the loop variable in the symbol table. If it does
+        # not exist then create it.
+        try:
+            loop_variable_symbol = symbol_table.lookup(loop_variable_name)
+        except KeyError:
+            # Add loop variable as it does not already exist
+            loop_variable_symbol = DataSymbol(loop_variable_name, INTEGER_TYPE)
+            symbol_table.add(loop_variable_symbol)
+
+        # Replace array access loop variable.
+        for array in assignment.walk(ArrayReference):
+            if not array.ancestor(ArrayReference):
+                # This is not a nested access e.g. a(b(n)).
+                array.children[array_index] = Reference(loop_variable_symbol)
+
+        # As this transformation affects ancestor nodes of the
```

Comment:
I think, as we discussed before, that this transformation should not affect ancestor nodes. i.e. it should simply replace e.g.

    my_array(ji+1,jtop,jk) = xxx

with

    do jj = jtop, jtop
      my_array(ji+1,jj,jk) = xxx
    end do

Moving the resulting loop should be done by a separate transformation if the user feels the need.

Reply:
This now hopefully better as we now place new loops innermost, but actually woult affect ancestor nodes when there is an existing NemoKern as we need to place the loop outside of that. I don't see a reasonable way round that, other than doing what we do now.